ldm/modules/lora_manager.py

```diff
@@ -0,0 +1,277 @@
+
+
+UNET_TARGET_REPLACE_MODULE = ["Transformer2DModel", "Attention"]
+TEXT_ENCODER_TARGET_REPLACE_MODULE = ["CLIPAttention", "CLIPMLP"]
+LORA_PREFIX_UNET = 'lora_unet'
+LORA_PREFIX_TEXT_ENCODER = 'lora_te'
+
+
+def lora_forward_hook(name):
+    def lora_forward(module, input_h, output):
+        if len(loaded_loras) == 0:
+            return output
+
+        for lora in applied_loras.values():
+            layer = lora.layers.get(name, None)
+            if layer is None:
+                continue
+            output = output + layer.up(layer.down(*input_h)) * lora.multiplier * layer.scale
+        return output
+
+    return lora_forward
+
+
+def load_lora(
+    name: str,
+    path_file: Path,
+    device: torch.device,
+    dtype: torch.dtype,
+    text_modules: dict[str, torch.nn.Module],
+    unet_modules: dict[str, torch.nn.Module],
+    multiplier=1.0
+):
+    print(f">> Loading lora {name} from {path_file}")
+    if path_file.suffix == '.safetensors':
+        checkpoint = load_file(path_file.absolute().as_posix(), device='cpu')
+    else:
+        checkpoint = torch.load(path_file, map_location='cpu')
+
+    lora = LoRA(name, multiplier)
+
+    alpha = None
+    rank = None
+    for key, value in checkpoint.items():
+        stem, leaf = key.split(".", 1)
+
+        if leaf.endswith("alpha"):
+            if alpha is None:
+                alpha = value.item()
+            continue
+
+        if stem.startswith(LORA_PREFIX_TEXT_ENCODER):
+            # text encoder layer
+            wrapped = text_modules.get(stem, None)
+            if wrapped is None:
+                print(f">> Missing layer: {stem}")
+                continue
+        elif stem.startswith(LORA_PREFIX_UNET):
+            # unet layer
+            wrapped = unet_modules.get(stem, None)
+            if wrapped is None:
+                print(f">> Missing layer: {stem}")
+                continue
+        else:
+            continue
+
+        if rank is None and leaf == 'lora_down.weight' and len(value.size()) == 2:
+            rank = value.shape[0]
+
+        if wrapped is None:
+            continue
+
+        layer = lora.layers.get(stem, None)
+        if layer is None:
+            layer = LoRALayer(name, stem, rank, alpha)
+            lora.layers[stem] = layer
+
+        if type(wrapped) == torch.nn.Linear:
+            module = torch.nn.Linear(
+                value.shape[1], value.shape[0], bias=False)
+        elif type(wrapped) == torch.nn.Conv2d:
+            module = torch.nn.Conv2d(
+                value.shape[1], value.shape[0], (1, 1), bias=False)
+        else:
+            print(
+                f">> Encountered unknown lora layer module in {name}: {type(value).__name__}")
+            continue
+
+        with torch.no_grad():
+            module.weight.copy_(value)
+
+        module.to(device=device, dtype=dtype)
+
+        if leaf == "lora_up.weight":
+            layer.up = module
+        elif leaf == "lora_down.weight":
```

Comment:
i'm a bit concerned about this - this is the sort of thing that i think should be avoided in invoke. given that huggingface diffusers is likely to support lora soon i would assume this will go away then? 

we had to do similar hacking to make `.swap()` work and i was very glad to pull it out in favour of the `CrossAttnProcessor` architecture in diffusers. in fact, isn't there already a `LoraCrossAttnProcessor`? and/or, can't this be better implemented as a `CrossAttnProcessor`?

Reply:
Yes, but `text_encoder` is not as straightforward. In any case, should probably change to something similar to the code here: https://github.com/huggingface/diffusers/pull/2141

Reply:
ahhh :(

Reply:
yeah the code linked seems like it might be a better solution. i'll resolve this conversation for now i guess.